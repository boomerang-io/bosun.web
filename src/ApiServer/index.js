import { Server, Serializer, Model } from "miragejs";
import { inflections } from "inflected";
import uuid from "uuid/v4";
import { serviceUrl } from "Config/servicesConfig";
import * as fixtures from "./fixtures";

export function startApiServer({ environment = "test", timing = 0 } = {}) {
  inflections("en", function (inflect) {
    // Prevent pluralization bc our apis are weird
    inflect.irregular("policy", "policy");
    inflect.irregular("violation", "violation");
    inflect.irregular("template", "template");
    inflect.irregular("insights", "insights");
  });

  return new Server({
    environment,
    // Load in mock data
    fixtures,
    // Return the data as is, don't add a root key
    serializers: {
      application: Serializer.extend({
        root: false,
        embed: true,
      }),
    },
    // Register the data as a model so we can use the schema
    models: {
      templates: Model,
      policies: Model,
    },

    routes() {
      // Control how long the responses take to resolve
      this.timing = timing;

      // Allow unhandled requests on the current domain to pass through
      this.passthrough();

      this.get("/info", () => []);

      /**
       * Simple GET of static data
       */
      this.get(serviceUrl.getUserProfile(), (schema) => {
        return schema.db.profile[0];
      });
      this.get(serviceUrl.getCicdNavigation({}), (schema, request) => {
        return schema.db.cicdNavigation.map((item) => {
          if (typeof item.link === "string") {
            item.link = item.link.replace(":teamId", request.queryParams.teamId);
          }

          return item;
        });
      });

      this.get(serviceUrl.getPlatformNavigation(), (schema) => {
        return schema.db.navigation[0];
      });

      this.get(serviceUrl.getTeams(), (schema) => {
        return schema.db.teams;
      });

      this.get(serviceUrl.getPolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        return activePolicy;
      });

      this.get(serviceUrl.getValidateInfo({ policyId: ":policyId" }), (schema) => {
        return schema.db.validateInfo;
      });

      this.get(serviceUrl.getTemplates(), (schema) => {
        return schema.db.templates;
      });

      this.get(serviceUrl.getPolicyOverview(), (schema, request) => {
        let { teamId } = request.queryParams;
        return schema.db.policies.filter((policy) => policy.teamId === teamId);
      });

      this.get(serviceUrl.getViolationsOverview(), (schema) => {
        return schema.db.violations;
      });

      this.get(serviceUrl.getInsightsOverview(), (schema) => {
        return schema.db.insights;
      });

      this.post(serviceUrl.postCreatePolicy(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.policies.create({ ...body, id: uuid() });
      });

      this.post(serviceUrl.postCreatePolicyTemplate(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.templates.create({ ...body, id: uuid() });
      });

      this.patch(serviceUrl.patchUpdatePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activePolicy = schema.policies.find(policyId);
        activePolicy.update(body);
        return schema.policies.all();
      });

      this.patch(serviceUrl.patchUpdatePolicyTemplate({ templateId: ":templateId" }), (schema, request) => {
        let { templateId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activeTemplate = schema.templates.find(templateId);
        activeTemplate.update(body);
        return schema.templates.all();
      });

      this.delete(serviceUrl.deletePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        activePolicy.destroy({ id: policyId });
        return schema.policies.all();
      });
    },
  });
}
