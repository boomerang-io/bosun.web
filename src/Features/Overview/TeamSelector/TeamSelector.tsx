import React from "react";
import { useMutation, queryCache } from "react-query";
import { useFeature } from "flagged";
import { Formik } from "formik";
import * as Yup from "yup";
import {
  ComboBox,
  TextInput,
  Button,
  ComposedModal,
  ModalFooter,
  ModalBody,
  ModalForm,
  Loading,
  InlineNotification,
  notify,
  ToastNotification,
} from "@boomerang-io/carbon-addons-boomerang-react";
import { FeatureFlag } from "Config/appConfig";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { Add16 } from "@carbon/icons-react";
import { PolicyTeam } from "Types";
import styles from "./TeamSelector.module.scss";

type Props = {
  activeTeam?: PolicyTeam | null;
  handleChangeTeam: (...args: any[]) => void;
  teams: PolicyTeam[];
};

export function TeamSelector({ activeTeam, handleChangeTeam, teams }: Props) {
  const [modalIsOpen, setModalIsOpen] = React.useState<boolean>(false);
  const [requestError, setRequestError] = React.useState<boolean>(false);
  const isStandaloneMode = useFeature(FeatureFlag.Standalone);
  const cancelRequestRef = React.useRef<any>();

  const [createTeamMutation, { isLoading: createIsLoading, error: createError }] = useMutation(
    (args: { body: { name: string } }) => {
      const { promise, cancel } = resolver.postCreateTeam(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryCache.invalidateQueries(serviceUrl.getTeams()),
    }
  );

  const selectedTeam = activeTeam
    ? {
        id: activeTeam.id,
        label: activeTeam.name,
        name: activeTeam.name,
      }
    : null;
  const teamsList = teams.map((team) => ({
    id: team.id,
    label: team.name,
    name: team.name,
  }));

  const teamNames = teams.map((team) => team.name);

  async function createTeam(body: { name: string }) {
    try {
      await createTeamMutation({ body });
      notify(<ToastNotification kind="success" title="Team Created" subtitle="Team successfully created" />);
      setModalIsOpen(false);
    } catch (e) {
      setRequestError(true);
      notify(<ToastNotification kind="error" title="Team Creation Failure" subtitle="Request to create team failed" />);
    }
  }

  return (
    <section className={styles.container}>
      <div className={styles.inputs}>
        <ComboBox
          id="team"
          onChange={handleChangeTeam}
          items={teamsList}
          labelText="Team"
          initialSelectedItem={selectedTeam}
          placeholder="Select a team"
          title="Team"
        />
        {isStandaloneMode ? (
          <Formik
            onSubmit={createTeam}
            initialValues={{ name: "" }}
            validateOnMount={false}
            validationSchema={Yup.object().shape({
              name: Yup.string().required("Enter a name").notOneOf(teamNames, "Name must be unique"),
            })}
            enableReinitialize
          >
            {(formikProps) => {
              const {
                isSubmitting,
                values,
                touched,
                errors,
                handleBlur,
                handleChange,
                handleSubmit,
                resetForm,
              } = formikProps;
              return (
                <>
                  <Button
                    renderIcon={Add16}
                    iconDescription="Add team"
                    onClick={() => setModalIsOpen(true)}
                    size="small"
                  >
                    Create Team
                  </Button>
                  <form onSubmit={handleSubmit}>
                    <ComposedModal
                      composedModalProps={{ containerClassName: styles.modalForm }}
                      modalHeaderProps={{
                        title: "Create team",
                        subtitle: "Add a new one",
                      }}
                      onCloseModal={() => {
                        resetForm();
                        setRequestError(false);
                        setModalIsOpen(false);
                      }}
                      isOpen={modalIsOpen}
                    >
                      {() => (
                        <ModalForm>
                          <ModalBody>
                            <TextInput
                              id="name"
                              placeholder="Enter team name"
                              labelText="Name"
                              onChange={handleChange}
                              onBlur={handleBlur}
                              invalid={errors.name && touched.name}
                              invalidText={errors.name}
                              value={values.name}
                            />
                            {requestError && (
                              <InlineNotification
                                lowContrast
                                kind="error"
                                title="Something's Wrong"
                                subtitle="Request to create version failed"
                                onCloseButtonClick={() => setRequestError(false)}
                              />
                            )}
                          </ModalBody>
                          {createIsLoading && <Loading />}
                          <ModalFooter>
                            <Button onClick={() => setModalIsOpen(false)} kind="secondary">
                              Cancel
                            </Button>
                            <Button
                              disabled={!values.name || Object.values(errors).length || isSubmitting}
                              type="submit"
                            >
                              {createIsLoading ? "Creating" : createError ? "Try Again" : "Create"}
                            </Button>
                          </ModalFooter>
                        </ModalForm>
                      )}
                    </ComposedModal>
                  </form>
                </>
              );
            }}
          </Formik>
        ) : null}
      </div>
    </section>
  );
}

export default TeamSelector;
