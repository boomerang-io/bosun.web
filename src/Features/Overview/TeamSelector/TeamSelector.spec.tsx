import React from "react";
import TeamSelector from ".";
import { teams } from "ApiServer/fixtures";
const mockfn = jest.fn();
describe("TeamSelector --- Snapshot", () => {
  it("Capturing Snapshot of TeamSelector", async () => {
    const { baseElement } = (global as any).rtlRender(<TeamSelector handleChangeTeam={mockfn} teams={teams} />);
    expect(baseElement).toMatchSnapshot();
  });
});
