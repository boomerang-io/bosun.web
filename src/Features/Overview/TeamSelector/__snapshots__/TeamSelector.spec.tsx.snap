// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamSelector --- Snapshot Capturing Snapshot of TeamSelector 1`] = `
<body
  id="app"
>
  <div>
    <section
      class="container"
    >
      <div
        class="inputs"
      >
        <div
          class="bx--bmrg-select"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="downshift-0-label"
            class="bx--list-box__wrapper"
            role="combobox"
          >
            <label
              class="bx--label"
              for="team"
              id="downshift-0-label"
            >
              <div
                style="display: flex;"
              >
                <div>
                  Team
                </div>
              </div>
            </label>
            <div
              aria-label="Choose an item"
              class="bx--combo-box bx--list-box"
            >
              <div
                aria-haspopup="true"
                aria-label="open menu"
                class="bx--list-box__field"
                data-toggle="true"
                id="team"
                role="button"
                tabindex="-1"
                type="button"
              >
                <input
                  aria-autocomplete="list"
                  aria-labelledby="downshift-0-label"
                  autocomplete="off"
                  class="bx--text-input bx--text-input--empty"
                  id="team"
                  placeholder="Select a team"
                  tabindex="0"
                  title="Team"
                  type="text"
                  value=""
                />
                
                <div
                  class="bx--list-box__menu-icon"
                >
                  <svg
                    aria-label="Open menu"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    name="chevron--down"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                    <title>
                      Open menu
                    </title>
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</body>
`;
